### ISSUE TITLE
Linter plugins: Add comments-related APIs

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#14564]
Comment-related APIs are not yet supported in JS plugins e.g. `context.sourceCode.getAllComments` (as noted in [docs](https://oxc.rs/docs/guide/usage/linter/js-plugins.html#api-support)).

The token-related APIs are going to be hard to implement, but the comment-related ones should not be too hard, as raw transfer can already deal with comments (implemented in `oxc-parser`).

#14488 demonstrates these APIs are required for `eslint-plugin-react-hooks` plugin.

**Comment by overlookmotel** [comment]
`getAllComments` has been added in #14589.

Next steps (stolen from @lilnasy's list on that PR):

- [X] ~~look into `RawTransferData { program.comments, comments }` duplication.~~ (no longer relevant)
- [X] oxlint's `interface Program` doesn't include the comments type. #14626
- [x] Lazy deserialization of `program.comments`
- [X] Remove `comment.parent` #14624
- [x] UTF8 to UTF16 conversion on the JS side
- [x] Test correct `start`/`end` offsets in files containing unicode chars
- [x] `getCommentsBefore`
- [x] `getCommentsAfter`
- [x] `getCommentsInside`
- [x] `commentsExistBetween`

**Comment by overlookmotel** [comment]
I'll tackle removing `comment.parent`. I tried just now, but it's trickier than I expected. The codegen needs some changes.

EDIT: Done in #14624.

**Comment by overlookmotel** [comment]
@lilnasy I've ticked a couple off the list. Do you have an interest in continuing work on this? (totally fine if you don't, just asking)

**Comment by lilnasy** [comment]
Yeah. I have a couple of branches I'm working on currently (see open pull requests for the ones ready to review). I'll get to comment APIs soon.

**Comment by overlookmotel** [comment]
For lazily deserializing comments:

Currently we have this in deserializer:

https://github.com/oxc-project/oxc/blob/9cd7a6777aef0a7816a372fdf4de88e2abdda8e7/apps/oxlint/src-js/generated/deserialize.js#L43-L77

I think we want to change:

```diff
- comments: deserializeVecComment(pos + 24), 
+ comments: null,
```

Then add somewhere below that in `deserializeProgram`:

```js
Object.defineProperty(program, 'comments', {
  get() {
    const comments = DESER[Vec<Comment>](POS_OFFSET.comments);
    Object.defineProperty(this, 'comments', { value: comments });
    return comments;
  },
});
```

Those changes can be made in the `raw_deser` attribute on `ProgramConverter` [here](https://github.com/oxc-project/oxc/blob/9cd7a6777aef0a7816a372fdf4de88e2abdda8e7/crates/oxc_ast/src/serialize/mod.rs#L123-L173).

(with the `Object.defineProperty` call gated with `if (COMMENTS)` so it only ends up in the linter deserializer)

**Comment by overlookmotel** [comment]
Oh no, there's a further complication.

Currently `deserializeWith` cleans up when it's done, setting `uint8` etc back to `undefined`, to allow the buffer to get garbage collected.

That's not going to work with lazy deserialization, because `deserializeVecComment` (which can get called later) needs the buffer (and `sourceText`) to read the comments from.

So, we'll need to:

* Alter `deserializeWith` so it doesn't clean up the buffer and source text.
* Add a new function `reset` which sets `unit8` etc to `undefined`.
* `deserialize` calls `reset` after calling `deserializeWith` (so that parser does same as it did before).
* Export `reset` from `deserialize.js`.
* In linter, call `reset` after it completes traversing the AST (at end of `lintFileImpl` [here](https://github.com/oxc-project/oxc/blob/9cd7a6777aef0a7816a372fdf4de88e2abdda8e7/apps/oxlint/src-js/plugins/lint.ts#L61-L165)).

**Comment by lilnasy** [comment]
- [x] Lazy deserialization of `program.comments`: #14637
- [x] UTF8 to UTF16 conversion on the JS side: #14768
- [x] Test correct `start`/`end` offsets in files containing unicode chars: #14768
- [x] `getCommentsBefore`: #14715
- [x] `getCommentsAfter`: #14715
- [x] `getCommentsInside`: #14715
- [x] `commentsExistBetween`: #14715
- [x] Make all comments APIs use binary search instead of linear search wherever possible: #14778

**Comment by camc314** [comment]
@lilnasy amazing work on this! Are all of these complete? or is there still work to do? 👀

**Comment by lilnasy** [comment]
Yes, everything discussed in this thread is done.

**Comment by camc314** [comment]
Amazing, thank you!
